File: 2018m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

class Contestant {
public:
    string name;
    int time;

    Contestant(const string& name, int time) : name(name), time(time) {}
};

int getIndexByName(const string& name, const vector<Contestant>& arr) {
    for (size_t i = 0; i < arr.size(); ++i) {
        if (arr[i].name == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

vector<Contestant> sortContestants(vector<Contestant> contestants) {
    size_t size = contestants.size();
    for (size_t i = 0; i < size; ++i) {
        for (size_t j = i + 1; j < size; ++j) {
            if (contestants[i].time > contestants[j].time ||
                (contestants[i].time == contestants[j].time &&
                 contestants[i].name > contestants[j].name)) {
                swap(contestants[i], contestants[j]);
            }
        }
    }
    return contestants;
}

void writeOutput(const vector<Contestant>& contestants) {
    ofstream output("U2rez.txt");

    for (const Contestant& contestant : contestants) {
        int time = contestant.time;
        int seconds = time % 60;
        int minutes = (time - seconds) / 60;
        output << contestant.name << string(20 - contestant.name.length(), ' ') << minutes << " " << seconds << "\n";
    }

    output.close();
}

int main() {
    ifstream data("U2.txt");

    vector<Contestant> contestants;
    vector<Contestant> finishedContestants;

    int n;
    data >> n;
    data.ignore();

    for (int i = 0; i < n; ++i) {
        string contestantData;
        getline(data, contestantData);
        string name = contestantData.substr(0, 20);
        contestantData = contestantData.substr(20);
        stringstream dataStream(contestantData);
        int hours, minutes, seconds;
        dataStream >> hours >> minutes >> seconds;
        int startTime = hours * 3600 + minutes * 60 + seconds;
        contestants.push_back(Contestant(name, startTime));
    }

    int m;
    data >> m;
    data.ignore();

    for (int i = 0; i < m; ++i) {
        string contestantData;
        getline(data, contestantData);
        string name = contestantData.substr(0, 20);
        contestantData = contestantData.substr(20);
        stringstream dataStream(contestantData);
        int hours, minutes, seconds;
        dataStream >> hours >> minutes >> seconds;
        int startTime = contestants[getIndexByName(name, contestants)].time;
        int endTime = hours * 3600 + minutes * 60 + seconds;
        finishedContestants.push_back(Contestant(name, endTime - startTime));
    }

    finishedContestants = sortContestants(finishedContestants);
    writeOutput(finishedContestants);

    data.close();

    return 0;
}

File: 2018m./uzd2./sprendimai/uzd2.py
class Contestant:
    def __init__(self, name, time):
        self.name = name
        self.time = time

def getIndexByName(name, arr):
    for i in range(len(arr)):
        if(arr[i].name == name):
            return i

def sortContestants(contestants):
    size = len(contestants)
    for i in range(size):
        for j in range(i + 1, size):
            if(contestants[i].time > contestants[j].time or 
                (contestants[i].time == contestants[j].time
                and contestants[i].name > contestants[j].name)):
                contestants[i], contestants[j] = contestants[j], contestants[i]
    return contestants

def writeOutput(contestants):
    output = open("U2rez.txt", "w")

    for i in range(len(contestants)):
        time = contestants[i].time
        seconds = time%60
        minutes = int((time-seconds)/60)
        output.write(contestants[i].name + " " * (20-len(contestants[i].name)) + str(minutes)+ " " + str(seconds) + "\n")

    output.close()

def main():
    data = open("U2.txt", "r")

    contestants = []
    finishedContestants = []

    for i in range(int(next(data))):
        contestantData = next(data)
        name = contestantData[:19]
        contestantData = contestantData[20:]
        contestantData = contestantData.split(" ")
        contestantData = [int(i) for i in contestantData]
        startTime = contestantData[0] * 3600 + contestantData[1] * 60 + contestantData[2]
        contestants.append(Contestant(name, startTime))

    for i in range(int(next(data))):
        contestantData = next(data)
        name = contestantData[:19]
        contestantData = contestantData[20:]
        contestantData = contestantData.split(" ")
        contestantData = [int(i) for i in contestantData]
        startTime = contestants[getIndexByName(name, contestants)].time
        endTime = contestantData[0] * 3600 + contestantData[1] * 60 + contestantData[2]
        finishedContestants.append(Contestant(name, endTime - startTime))
    
    finishedContestants = sortContestants(finishedContestants)

    writeOutput(finishedContestants)

    data.close()

if __name__ == "__main__":
    main()